src/machine/machine_mimxrt1062.go

```diff
@@ -879,3 +880,190 @@ func (p Pin) getMuxMode(config PinConfig) uint32 {
+		} else if samples >= 8 {
+			// 8 samples averaged
+			mode |= (nxp.ADC_CFG_AVGS_AVGS_1 << nxp.ADC_CFG_AVGS_Pos) & nxp.ADC_CFG_AVGS_Msk
+		} else {
+			// 4 samples averaged
+			mode |= (nxp.ADC_CFG_AVGS_AVGS_0 << nxp.ADC_CFG_AVGS_Pos) & nxp.ADC_CFG_AVGS_Msk
+		}
+		average = nxp.ADC_GC_AVGE
+	}
+
+	return mode, average
+}
+
+const noChannel = uint8(0xFF)
+
+// getChannel returns the input channel for ADC1 & ADC2 of the receiver Pin p.
+func (p Pin) getChannel() (adc1, adc2 uint8, ok bool) {
+	switch p.getPort() {
+	case portA:
+		switch p.getPos() {
+		case 12: // [AD_B0_12]:       ADC1_IN1        ~
+			return 1, noChannel, true
+		case 13: // [AD_B0_13]:       ADC1_IN2        ~
+			return 2, noChannel, true
+		case 14: // [AD_B0_14]:       ADC1_IN3        ~
+			return 3, noChannel, true
+		case 15: // [AD_B0_15]:       ADC1_IN4        ~
+			return 4, noChannel, true
+		case 16: // [AD_B1_00]:       ADC1_IN5     ADC2_IN5
+			return 5, 5, true
+		case 17: // [AD_B1_01]:       ADC1_IN6     ADC2_IN6
+			return 6, 6, true
+		case 18: // [AD_B1_02]:       ADC1_IN7     ADC2_IN7
+			return 7, 7, true
+		case 19: // [AD_B1_03]:       ADC1_IN8     ADC2_IN8
+			return 8, 8, true
+		case 20: // [AD_B1_04]:       ADC1_IN9     ADC2_IN9
+			return 9, 9, true
+		case 21: // [AD_B1_05]:       ADC1_IN10    ADC2_IN10
+			return 10, 10, true
+		case 22: // [AD_B1_06]:       ADC1_IN11    ADC2_IN11
+			return 11, 11, true
+		case 23: // [AD_B1_07]:       ADC1_IN12    ADC2_IN12
+			return 12, 12, true
+		case 24: // [AD_B1_08]:       ADC1_IN13    ADC2_IN13
+			return 13, 13, true
+		case 25: // [AD_B1_09]:       ADC1_IN14    ADC2_IN14
+			return 14, 14, true
+		case 26: // [AD_B1_10]:       ADC1_IN15    ADC2_IN15
+			return 15, 15, true
+		case 27: // [AD_B1_11]:       ADC1_IN0     ADC2_IN0
+			return 16, 16, true
+		case 28: // [AD_B1_12]:          ~         ADC2_IN1
+			return noChannel, 1, true
+		case 29: // [AD_B1_13]:          ~         ADC2_IN2
+			return noChannel, 2, true
+		case 30: // [AD_B1_14]:          ~         ADC2_IN3
+			return noChannel, 3, true
+		case 31: // [AD_B1_15]:          ~         ADC2_IN4
+			return noChannel, 4, true
+		default:
+			return noChannel, noChannel, false
+		}
+	default:
+		return noChannel, noChannel, false
+	}
+}
+
+type adcStatus struct{ volatile.Register32 }
+
+var adcStatusGlobal adcStatus
+
+func (s adcStatus) isCalibrating() bool { return s.Get() != 0 }
+func (s adcStatus) startCalibrating()   { s.Set(1) }
+func (s adcStatus) stopCalibrating()    { s.Set(0) }
+func (s adcStatus) waitForCalibration() {
+	if s.isCalibrating() {
+		work := sync.WaitGroup{}
+		work.Add(2)
+		go func(w *sync.WaitGroup) {
+			for nxp.ADC1.GC.HasBits(nxp.ADC_GC_CAL) {
+			} // wait for calibration
+			w.Done()
+		}(&work)
+		go func(w *sync.WaitGroup) {
+			for nxp.ADC2.GC.HasBits(nxp.ADC_GC_CAL) {
+			} // wait for calibration
+			w.Done()
+		}(&work)
+		work.Wait()
```

Comment:
Why not keep it simple? Goroutines won't help here as they're busy-waiting anyway (and I'd like to avoid expensive operations such as goroutines in the machine package).

```suggestion
		for nxp.ADC1.GC.HasBits(nxp.ADC_GC_CAL) {
		} // wait for calibration
		for nxp.ADC2.GC.HasBits(nxp.ADC_GC_CAL) {
		} // wait for calibration
```

Reply:
Replaced goroutines with simpler spin locks as suggested